### PR TITLE
Feature/caas metadata

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

- Upgrade `charm.v6`;

- Add `MinVersion`, `Daemonset` support in Deployment in `metadata.yaml`;

## QA steps

- deploy caas charm with the new format of deployment in metadata.yaml;

```yaml
deployment:
  type: stateful | stateless | daemon
  service: omit  # new value to omit a `service`;
  min-version: "1.10.1"  # new value to limit the min version of k8s cluster for the charm;
```

## Documentation changes

Yes

## Bug reference

None

### CODE REVIEW

File: apiserver/facades/client/application/application_unit_test.go [hunk 1]
```diff
@@ -701,6 +702,42 @@ func (s *ApplicationSuite) TestDeployAttachStorage(c *gc.C) {
 	c.Assert(results.Results[2].Error, gc.ErrorMatches, `"volume-baz-0" is not a valid volume tag`)
 }
 
+func (s *ApplicationSuite) TestDeployMinDeploymentVersionTooHigh(c *gc.C) {
+	s.model.modelType = state.ModelTypeCAAS
+	s.backend.charm = &mockCharm{
+		meta: &charm.Meta{
+			Deployment: &charm.Deployment{
+				DeploymentType: "stateless",
```
Comment (wallyworld) [inline apiserver/facades/client/application/application_unit_test.go]: This test is for min version check so we don't need to set these other unrelated deployment attributes

File: apiserver/facades/client/application/charmstore.go [hunk 2]
```diff
@@ -248,11 +249,32 @@ func openCSClient(csAPIURL string, args params.AddCharmWithAuthorization) (*cscl
...
+		// append build number if it's not specified.
+		charmDeployment.MinVersion += ".0"
+	}
+	minver, err = version.Parse(charmDeployment.MinVersion)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if minver != version.Zero && minver.Compare(*caasVersion) > 0 {
+		return errors.NewNotValid(nil, fmt.Sprintf(
+			"charm's deployment min version(%q) is higher than remote cloud version (%q)",
```
Comment (wallyworld) [inline apiserver/facades/client/application/charmstore.go]: This method is only intended to check the juju version and adding the k8s version check makes it bloated and a bit messy. 
We really want a separate checkMinK8sVersion() method.

The k8s error message should be something like:
"charm requires a minimum k8s version of %v but the cluster only runs version %v"

File: caas/broker.go [hunk 3]
```diff
@@ -81,16 +81,24 @@ const (
...
 
+// IsOmit indicates if a service is required.
+func (st ServiceType) IsOmit() bool {
+	return st == ServiceOmit
+}
+
 const (
 	ServiceCluster      ServiceType = "cluster"
 	ServiceLoadBalancer ServiceType = "loadbalancer"
 	ServiceExternal     ServiceType = "external"
```
Comment (wallyworld) [inline caas/broker.go]: did we include "external" in the charm.v6 code?

File: caas/kubernetes/provider/k8s.go [hunk 4]
```diff
@@ -1781,6 +1789,9 @@ func (k *kubernetesClient) configureService(
 			serviceType = core.ServiceTypeLoadBalancer
 		case caas.ServiceExternal:
 			serviceType = core.ServiceTypeExternalName
+		case caas.ServiceOmit:
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: this conflicts with "if hasService" above

File: caas/broker.go (hunk 3 above)
Comment (ycliuhw) [reply caas/broker.go]: yes, we did. https://github.com/juju/charm/blob/v6/meta.go#L166

File: caas/kubernetes/provider/k8s.go (hunk 4 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go]: ah, right, this case is unreachable. removing

File: apiserver/facades/client/application/charmstore.go [hunk 5]
```diff
@@ -249,13 +249,15 @@ func openCSClient(csAPIURL string, args params.AddCharmWithAuthorization) (*cscl
...
-	// check Juju min version.
+func checkJujuMinVersion(ch charm.Charm) (err error) {
 	minver := ch.Meta().MinJujuVersion
 	if minver != version.Zero && minver.Compare(jujuversion.Current) > 0 {
 		return minVersionError(minver, jujuversion.Current)
 	}
+	return nil
+}
 
+func checkCaaSMinVersion(ch charm.Charm, caasVersion *version.Number) (err error) {
```
Comment (hpidcock) [inline apiserver/facades/client/application/charmstore.go]: checkCAASMinVersion

File: api/caasunitprovisioner/client.go [hunk 6]
```diff
@@ -147,6 +147,7 @@ func (c *Client) WatchPodSpec(application string) (watcher.NotifyWatcher, error)
 type DeploymentInfo struct {
 	DeploymentType string
 	ServiceType    string
+	Daemonset      bool
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply api/caasunitprovisioner/client.go]: changed the Daemonset a deployment type now;

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
FAIL: flag_test.go:94: WorkerSuite.TestErrRefresh
non-related failure.
$$merge$$

**Pull request merged**